### ISSUE TITLE
Continue learning not working as expected with chunked files 

### ISSUE THREAD

**Issue by yairdata** [microsoft/LightGBM#3194]
i am using the python package on CPU, python 3.6
lightgbm version 2.3.0
since i am running with limited resources on a single machine and read a large file , i am using chunked option when reading the pandas.
i expected that the log loss will decrease over time but with each new chunk on the input data - the logloss doesn't fit the best iteration of the previous learning.

```python
 incremental_dataframe = pd.read_csv("myFile.csv",  chunksize=100000,low_memory=False)
###some pre-process code
my_iter=0
for df in incremental_dataframe:
    my_iter+=1
    d_train = lgb.Dataset(X_train[cat_cols_try+num_cols_try],
                      label=y_train,
                      free_raw_data=False)
    if (my_iter==1):
        lgbModel = lgb.train(lgb_params,
                            categorical_feature=cat_cols_try,
                             train_set=d_train,
                             valid_sets=d_val)
        lgbModel.save_model('model.txt', num_iteration=lgbModel.best_iteration)
    else:
        print("next Iter")
        lgb_estimator = lgb.train(lgb_params,
                             categorical_feature=cat_cols_try,
                             # Pass partially trained model:
                             init_model='model.txt',
                         keep_training_booster=True,
                             train_set=d_train,
                             valid_sets=d_val)
        lgb_estimator.save_model("model.txt", num_iteration=lgb_estimator.best_iteration)
```

relevant log section:

```
[3909]	valid_0's binary_logloss: 0.00470439
[3910]	valid_0's binary_logloss: 0.00470428
[3911]	valid_0's binary_logloss: 0.00470451
[3912]	valid_0's binary_logloss: 0.00470464
Early stopping, best iteration is:
[3862]	valid_0's binary_logloss: 0.00470362
next Iter
[3863]	valid_0's binary_logloss: 0.00480829
Training until validation scores don't improve for 50 rounds
[3864]	valid_0's binary_logloss: 0.00480429
[3865]	valid_0's binary_logloss: 0.00480014
[3866]	valid_0's binary_logloss: 0.00479618
```

when i tried it on a single file with one pass , with the train method the results were as expected.

**Comment by jameslamb** [comment]
Hi @yairdata , thanks for using `LightGBM`! I've reformatted your description (hope you don't mind) so it is easier to read.

In case are you are new to markdown, you should see [this documentation from GitHub](https://help.github.com/en/github/writing-on-github/basic-writing-and-formatting-syntax#quoting-code).

This is a cool use of chunked data frames + training continuation!

Can you please clarify what you expected the behavior to be here and how LightGBM's behavior differs from it? I didn't quite understand this:

> the logloss doesn't fit the best iteration of the previous learning.

In general, if each new iteration is also a new chunk of data, I don't think it's safe to assume that the loss will continue to decrease. If the dataset is sorted in some way, a model fit to the first 5 chunks might not describe the 6th very well. `pandas` chunked data frame returns an iterator over _consecutive, non-overlapping_ chunks, so it is different than training on batches of data drawn randomly from the same dataset at each iteration.

**Comment by yairdata** [comment]
thanks for the answer @jameslamb !
i didn't quite get the point of chunked vs. randomly drawn data from same dataset.
isn't it essentially the same - new data is encountered by the model when it tries to keep training?
what i am looking for is a way to deal with data that is larger than memory on a single machine , so i can't load all of it into a pandas dataframe.
i thought i can use the continue training technique to do so.
if it can or there is another way of handling such a scenario - please share.

**Comment by guolinke** [comment]
@yairdata 

I agree with @jameslamb , I don't the continued training will always improve the performance on the fixed validation set. 
For the large dataset, you can set the data to csv file first, then pass the filename to lgb.Dataset constructor, with additional parameter `two_round=True`.

**Comment by yairdata** [comment]
thanks @guolinke .
can you provide short code example of that technique?

**Comment by guolinke** [comment]
it is like
```python
d_train = lgb.Dataset('path_to_your_file',params={'two_round': True})
```
